# Importing libraries

In [4]:
#import pandas numpy and os
import pandas as pd
import numpy as np
import os

In [2]:
#creating path
pathk=r'/Users/keerthiputti/Desktop/Data Analysis/Instacart Basket Analysis-Python'

In [6]:
df_prod=pd.read_csv(os.path.join(pathk,'02 Data','Original Data','products.csv'),index_col=False)

In [7]:
dfords=pd.read_csv(os.path.join(pathk,'02 Data','Prepared Data','orders_wrangled.csv'),index_col=False)

In [9]:
df_prod

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [10]:
dfords

Unnamed: 0  order_id  user_id  order_number  Order_day_of_the_week  \
0                 0   2539329        1             1                      2   
1                 1   2398795        1             2                      3   
2                 2    473747        1             3                      3   
3                 3   2254736        1             4                      4   
4                 4    431534        1             5                      4   
...             ...       ...      ...           ...                    ...   
3421078     3421078   2266710   206209            10                      5   
3421079     3421079   1854736   206209            11                      4   
3421080     3421080    626363   206209            12                      1   
3421081     3421081   2977660   206209            13                      1   
3421082     3421082    272231   206209            14                      6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

Practice work

# Mixed value checks

In [11]:
#creating data frame
df_test=pd.DataFrame()

In [12]:
#adding values to data frame
df_test['mix']=['a','b',1,True]

In [13]:
#display
df_test.head()

mix
0     a
1     b
2     1
3  True

In [ ]:
checking mixed datatypes in test dataframe

In [14]:
#checking for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)


mix


In [ ]:
checking mixed datatypes in products dataframe

In [15]:
for col in df_prod.columns.tolist():
  weird = (df_prod[[col]].applymap(type) != df_prod[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prod[weird]) > 0:
    print (col)

product_name


checking mixed datatypes in order dataframe

In [16]:
for col in dfords.columns.tolist():
  weird = (dfords[[col]].applymap(type) != dfords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (dfords[weird]) > 0:
    print (col)

# Missing value checks

In [18]:
#checking missing values in prod
df_prod.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [22]:
#creating subset data frame for null values in product_name column

df_nan=df_prod[df_prod['product_name'].isnull()==True]

In [63]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [65]:
#checking shape of df_prod
df_prod.shape

(49693, 5)

In [66]:
#dropping null value rows

df_prod_clean = df_prod[df_prod['product_name'].isnull() == False]

In [67]:
df_prod_clean.shape

(49677, 5)

In [68]:
#another way to drop na rows

df_prod.dropna(inplace = True)

In [69]:
df_prod.shape

(49677, 5)

In [71]:
#create subset of duplicate rows

df_dups = df_prod_clean[df_prod_clean.duplicated()]

In [74]:
df_dups.head()

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [75]:
#removing duplicates and moving a new data frame

df_prod_clean_no_dup=df_prod_clean.drop_duplicates()

In [77]:
df_prod_clean_no_dup.shape

(49672, 5)

In [78]:
df_prod_clean_no_dup.to_csv(os.path.join(pathk,'02 Data','Prepared Data','products_checked.csv'))

End of practice codes

# Begining of task 4.5 

In [80]:
#4(2) checking any inconsistencies
df_prod.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

I find the max price 99999.00 looks out of place.Instacart sells household goods and the max price seems really large amount

In [81]:
#4(3) checking for mixed data type in orders data
df_ords=pd.read_csv(os.path.join(pathk,'02 Data','Original Data','Orders.csv'),index_col=False)

In [82]:
# mixed data type code, running for df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No output on running the mixed data type code. meaning there is no column with mixed data type in order.csv
hence nothing to do for 4(4)

In [83]:
# 4(5) missing value check
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The column "day_since_prior_order" has 206209 entries of null values. I believe we will not be able to remove this, as this is a huge number of records. we should either replace null values with mean or median as the column is integer value.I will run describe the data frame and see which method will be better.

In [85]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

It will be better to impute the data with median as it fits the column best. Mean will be in decimal and column takes integer values.

In [97]:
df_ords['days_since_prior_order'].median()

7.0

In [98]:
# 4(6) replacing null values with median
df_ords['days_since_prior_order'].fillna(7.0, inplace=True)

In [99]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.421083e+06  
mean        1.345202e+01            1.086681e+01  
std         4.226088e+00            8.978521e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            5.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [100]:
#4(7) checking for duplicates
df_dup_ords=df_ords[df_ords.duplicated()]

In [101]:
#checking duplicates
df_dup_ords

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

4(8) 0 rows, hence no duplciates found in order.csv

In [103]:
#exporting cleaned products data to csv
df_prod_clean_no_dup.to_csv(os.path.join(pathk,'02 Data','Prepared Data','products_cleaned.csv'))

In [105]:
#exporting cleaned orders data to csv
df_ords.to_csv(os.path.join(pathk,'02 Data','Prepared Data','orders_cleaned.csv'))

In [106]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [107]:
df_ords_cleaned=df_ords.drop(columns=['eval_set'])

In [108]:
df_ords_cleaned

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           7.0  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [110]:
df_ords_cleaned['order_id']=df_ords_cleaned['order_id'].astype('str')

In [111]:
df_ords_cleaned['user_id']= df_ords_cleaned['user_id'].astype('str')

In [112]:
df_ords_cleaned.head()

order_id user_id  order_number  order_dow  order_hour_of_day  \
0  2539329       1             1          2                  8   
1  2398795       1             2          3                  7   
2   473747       1             3          3                 12   
3  2254736       1             4          4                  7   
4   431534       1             5          4                 15   

   days_since_prior_order  
0                     7.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [113]:
df_ords_cleaned.rename(columns={"order_dow":"Order_day_of_the_week"},inplace=True)

In [114]:
df_ords_cleaned.to_csv(os.path.join(pathk,'02 Data','Prepared Data','orders_clean_agn.csv'))